app/controllers/links_controller.rb

```diff
@@ -59,6 +64,17 @@ def new
     @link.expires = Time.now.utc + 1.days
   end
 
+  # GET /links/1/history
+  def history
+    @past_links = PastLink.all.order(id: :desc).where(link: @link).take(50)
+  end
+
+  # POST /links/:id/history/:past_link_id/reaction
+  def new_reaction
+    @past_link.set_reaction(past_link_params['response_type'], past_link_params['response_text'])
+    redirect_to url_for(controller: :links, action: :new_reaction, id: @link.id)
```

Comment:
todo: return error page if setting the reaction fails

not too concerned because you can only send invalid data here if you tamper with the page